### ISSUE TITLE
Global Loot Modifier cannot use datapack registries (like enchantments)

### ISSUE THREAD

**Issue by HyCraftHD** [MinecraftForge/MinecraftForge#10042]
**Minecraft Version:** 1.21

**Forge Version:** 51.0.23

**Logs:** ```Could not decode GlobalLootModifier with json id uteamtest_multiloader:test_enchantment_modifier - error: Not a json array: "uteamtest_multiloader:test" missed input: {"minecraft:enchantments":[{"enchantments":"uteamtest_multiloader:test","levels":{"min":1}}]}```

**Steps to Reproduce:**
 1. Add a GLM with an enchantment
```js
 {
	"type": "uteamtest_multiloader:test",
	"conditions": [
		{
			"condition": "minecraft:match_tool",
			"predicate": {
				"predicates": {
					"minecraft:enchantments": [
						{
							"enchantments": "uteamtest_multiloader:test",
							"levels": {
								"min": 1
							}
						}
					]
				}
			}
		}
	]
}
```
 2. Loot modifier is not loaded correctly because registry lookup is not provided

A fix would be to pass in the correct Ops with the Registry Context:
https://github.com/MinecraftForge/MinecraftForge/blob/e2d1ca7d0ead9c59173c5cc089ea0fe99ec87532/src/main/java/net/minecraftforge/common/loot/LootModifierManager.java#L72
```JsonOps.INSTANCE``` should be changed to ```registries.createSerializationContext(JsonOps.INSTANCE)```.
One problem is that the registration context does not seem to be directly available here.

**Comment by LexManos** [comment]
The error states that your decoder is expecting an array. Says nothing about custom registies in the error.
Post the actual full logs and your full decoder code.

**Comment by HyCraftHD** [comment]
Thats unfortunately the whole error message. There are no custom decoders used.
Here is the full log: https://gist.github.com/HyCraftHD/92a65cdf5d061310e23d1f33b51227f8

I have tested it with providing the registry access with the test mods in forge and the same issue is fixed with this. (PR #10043)

Before:
```java
[18:28:18] [Render thread/WARN] [ne.mi.co.lo.LootModifierManager/]: Could not decode GlobalLootModifier with json id global_loot_test:bamboo_silk - error: Not a json array: "minecraft:bamboo"
[m[33m[18:28:18] [Render thread/WARN] [ne.mi.co.lo.LootModifierManager/]: Could not decode GlobalLootModifier with json id global_loot_test:wheat_harvest - error: Not a json array: "minecraft:shears"
[m[33m[18:28:18] [Render thread/WARN] [ne.mi.co.lo.LootModifierManager/]: Could not decode GlobalLootModifier with json id global_loot_test:smelting - error: Not a json array: "global_loot_test:smelt" missed input: {"minecraft:enchantments":[{"enchantments":"global_loot_test:smelt","levels":{"min":1}}]}
```

After: Working and no error.

**Comment by LexManos** [comment]
Okay so that took some looking into. But I figured it out.
HolderSetCodec has two paths, one for when you have the registry context, and one for when you don't.
The second path only work when using Direct holder sets. It'll error with can't access registry in other cases.
However, here's the kicker, the vanilla code uses a different codec for encoding then it does for decoding. Which causes it to fail with the uninformative `error: Not a json array: "minecraft:shears";`

I have now fixed that to use the same codec on both ends, which means the error is the much more informative:
`error: Failed to parse either. First: Not a json array: "minecraft:shears"; Second: Can't access registry ResourceKey[minecraft:root / minecraft:item]`

On top of that i've made the game tests for this more robust, so that it properly catches this case in the future to make sure it doesn't happen again.

https://github.com/MinecraftForge/MinecraftForge/commit/d498ec4a4580d6a806cbaea071258b74af1ebeb7
